src/components/metadataFormControls/ModelTransfer/BaseModelTransfer.tsx

```diff
@@ -0,0 +1,67 @@
+import { Transfer, TransferProps } from '@dhis2/ui'
+import React, { useCallback, useMemo } from 'react'
+import { DisplayableModel } from '../../../types/models'
+
+const toDisplayOption = (model: DisplayableModel) => ({
+    value: model.id,
+    label: model.displayName,
+})
+
+type OwnProps<TModel> = {
+    selected: TModel[]
+    available: TModel[]
+    onChange: ({ selected }: { selected: TModel[] }) => void
+}
+
+export type BaseModelTransferProps<TModel> = Omit<
+    TransferProps,
+    keyof OwnProps<TModel> | 'options' | 'selected'
+> &
+    OwnProps<TModel>
+
+/* Simple wrapper component handle generic models with Transfer-component. */
+export const BaseModelTransfer = <TModel extends DisplayableModel>({
+    available,
+    selected,
+    onChange,
+    ...transferProps
+}: BaseModelTransferProps<TModel>) => {
+    const { allModelsMap, allTransferOptions } = useMemo(() => {
+        const allModels = selected.concat(available)
+        const allModelsMap = new Map(allModels.map((o) => [o.id, o]))
+        const allTransferOptions = allModels.map(toDisplayOption)
+        return {
+            allModelsMap,
+            allTransferOptions,
+        }
+    }, [available, selected])
+
+    const selectedTransferValues = useMemo(
+        () => selected.map((s) => s.id),
+        [selected]
+    )
+
+    const handleOnChange = useCallback(
+        ({ selected }: { selected: string[] }) => {
+            // map the selected ids to the full model
+            // loop through selected to keep order
+            const selectedModels = selected
+                .map((id) => allModelsMap.get(id))
+                .filter((model) => !!model)
```

Comment:
You referring to the filter? 
You're right, but we would have to do a type assertion and this was mostly to please the compiler, because .get() on a map always include undefined.